### PR TITLE
RD-14190-Redis-and-Mongo-reduced-instrumentation

### PR DESCRIPTION
**Pull request by eugene-lumigo**

### CODE REVIEW

File: src/samplers/mongodbSampler.test.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+      'spanName',
+      SpanKind.CLIENT,
+      {},
+      []
+    );
+    expect(result.decision).toBe(SamplingDecision.RECORD_AND_SAMPLED);
+  });
+
+  it('should return NOT_RECORD when dbSystem is mongodb and dbOperation is isMaster and LUMIGO_REDUCED_MONGO_INSTRUMENTATION is true', () => {
+    process.env.LUMIGO_REDUCED_MONGO_INSTRUMENTATION = 'true';
```
Comment (harelmo-lumigo) [inline src/samplers/mongodbSampler.test.ts:26]: In general, I think it's better to split the tests where `process.env.LUMIGO_REDUCED_MONGO_INSTRUMENTATION` is `true` and `false`, and put those under two separate `describe` blocks, like this:
```ts
describe("when LUMIGO_REDUCED_MONGO_INSTRUMENTATION is 'true', () => {
    beforeEach(() => {
        process.env.LUMIGO_REDUCED_MONGO_INSTRUMENTATION = 'true';
    })

    // tests for `true`
})

describe("when LUMIGO_REDUCED_MONGO_INSTRUMENTATION is 'false', () => {
    beforeEach(() => {
        process.env.LUMIGO_REDUCED_MONGO_INSTRUMENTATION = 'false';
    })

    // tests for `false`
})
```

File: src/samplers/mongodbSampler.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    context: Context,
+    traceId: string,
+    spanName: string,
+    spanKind: SpanKind,
+    attributes: Attributes,
+    links: Link[]
+  ): SamplingResult {
+    // Note, there is probably a bug in opentelemetry api, making mongoSampler always receives attributes array empty.
+    // This makes it impossible to filter based on db.system and db.operation attributes. Filter based on spanName only.
+    // Opentemetry version upgrade might fix this issue.
```
Comment (harelmo-lumigo) [inline src/samplers/mongodbSampler.ts:22]: Are we on the latest version, and it happens with it?

File: src/samplers/mongodbSampler.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    // https://lumigo.atlassian.net/browse/RD-14250
+    let decision = SamplingDecision.RECORD_AND_SAMPLED;
```
Comment (harelmo-lumigo) [inline src/samplers/mongodbSampler.ts:24]: I'm usually refraining from using `let`, as IMO it's less easy to follow and les explicit. WDYT about:
```ts
if (spanKind === SpanKind.CLIENT && matchMongoIsMaster(spanName, dbSystem, dbOperation)) {
      logger.debug(
        `Drop span ${spanName} with db.system: ${dbSystem} and db.operation: ${dbOperation}, because LUMIGO_REDUCED_MONGO_INSTRUMENTATION is enabled`
      );
      return SamplingDecision.NOT_RECORD;
    }
```

and then the fallback `return` becomes:
```ts
return { decision: SamplingDecision.RECORD_AND_SAMPLED; };
```

File: README.md [hunk 4]
```diff
@@ -122,6 +122,8 @@ This setting is independent from `LUMIGO_DEBUG`, that is, `LUMIGO_DEBUG` does no
...
   * `LUMIGO_FILTER_HTTP_ENDPOINTS_REGEX_SERVER` applies the regular expression search exclusively to server spans. Searching is performed against the following attributes on a span: `url.path` and `http.target`.
   * `LUMIGO_FILTER_HTTP_ENDPOINTS_REGEX_CLIENT` applies the regular expression search exclusively to client spans. Searching is performed against the following attributes on a span: `url.full` and `http.url`.
+* `LUMIGO_REDUCED_MONGO_INSTRUMENTATION=true`: Reduces the amount of data collected by the MongoDB [instrumentation](https://www.npmjs.com/package/@opentelemetry/instrumentation-mongodb), such as not collecting the `db.operation` attribute `isMaster`. By default, the MongoDB instrumentation reduces the amount of data collected.
+* `LUMIGO_REDUCED_REDIS_INSTRUMENTATION=true`: Reduces the amount of data collected by the Redis [instrumentation](https://www.npmjs.com/package/@opentelemetry/instrumentation-redis-4), such as not collecting the `db.statement` attribute `INFO`. By default, the Redis instrumentation reduces the amount of data collected.
```
Comment (harelmo-lumigo) [inline README.md:126]: After going through the PR it wasn't clear that it's the new default - I'd rephrase it a bit to emphasize that.
```suggestion
* `LUMIGO_REDUCED_MONGO_INSTRUMENTATION=true`: Reduces the amount of data collected by the MongoDB [instrumentation](https://www.npmjs.com/package/@opentelemetry/instrumentation-mongodb), such as not collecting the `db.operation` attribute `isMaster`. 
Defaults to `true`, meaning the MongoDB instrumentation reduces the amount of data collected.
* `LUMIGO_REDUCED_REDIS_INSTRUMENTATION=true`: Reduces the amount of data collected by the Redis [instrumentation](https://www.npmjs.com/package/@opentelemetry/instrumentation-redis-4), such as not collecting the `db.statement` attribute `INFO`. 
Defaults to `true`, meaning the Redis instrumentation reduces the amount of data collected.
```

Also, is product aware of that change of defaults? I guess it might break or confuse existing users relying on the current data collected...

File: src/samplers/mongodbSampler.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+): boolean => {
+  const reduceMongoInstrumentation = process.env.LUMIGO_REDUCED_MONGO_INSTRUMENTATION;
+  let isReducedMongoInstrumentationEnabled: boolean;
+
+  if (reduceMongoInstrumentation == null || reduceMongoInstrumentation === '') {
+    isReducedMongoInstrumentationEnabled = true; // Default to true
+  } else if (reduceMongoInstrumentation.toLowerCase() === 'true') {
+    isReducedMongoInstrumentationEnabled = true;
+  } else
+    isReducedMongoInstrumentationEnabled = reduceMongoInstrumentation.toLowerCase() !== 'false';
```
Comment (harelmo-lumigo) [inline src/samplers/mongodbSampler.ts:65]:
```suggestion
const reduceMongoInstrumentation = process.env.LUMIGO_REDUCED_MONGO_INSTRUMENTATION;
const isReducedMongoInstrumentationEnabled =
  reduceMongoInstrumentation == null || reduceMongoInstrumentation === '' || reduceMongoInstrumentation.toLowerCase() !== 'false';
```

File: test/instrumentations/mongodb/mongodb.test.ts [hunk 6]
```diff
@@ -205,5 +210,76 @@ describe.each(versionsToTest(INSTRUMENTATION_NAME, INSTRUMENTATION_NAME))(
...
+
+          await testApp.invokeGetPath(`/mongodb-isMaster`);
+          // older versions of mongodb driver add extra spans
+          let spans = await testApp.getFinalSpans(2);
+          expect(getSpansByAttribute(spans, 'db.system', 'mongodb')).toHaveLength(1);
+        } else {
+          await testApp.invokeGetPath(`/mongodb-isMaster`);
+          let spans = await testApp.getFinalSpans(1);
+          expect(getSpansByAttribute(spans, 'db.system', 'mongodb')).toHaveLength(0);
+        }
```
Comment (harelmo-lumigo) [inline test/instrumentations/mongodb/mongodb.test.ts:245]: I usually try to avoid branching in tests. WDYT about this structure? 
```suggestion
        await testApp.invokeGetPath(`/mongodb-isMaster`);
        const expectedSpanCount = versionToTest.startsWith('3') ? 2 : 1;
        const expectedDbSystemAttributeSpans = versionToTest.startsWith('3') ? 1 : 0;
        
        const spans = await testApp.getFinalSpans(expectedSpanCount);
        expect(getSpansByAttribute(spans, 'db.system', 'mongodb')).toHaveLength(expectedDbSystemAttributeSpans);
```

File: test/instrumentations/mongodb/mongodb.test.ts [hunk 7]
```diff
@@ -205,5 +210,76 @@ describe.each(versionsToTest(INSTRUMENTATION_NAME, INSTRUMENTATION_NAME))(
...
+
+          await testApp.invokeGetPath(`/mongodb-isMaster`);
+          // older versions of mongodb driver add extra spans
+          let spans = await testApp.getFinalSpans(3);
+          expect(getSpansByAttribute(spans, 'db.operation', 'isMaster')).toHaveLength(1);
+        } else {
+          await testApp.invokeGetPath(`/mongodb-isMaster`);
+          let spans = await testApp.getFinalSpans(2);
+          expect(getSpansByAttribute(spans, 'db.operation', 'isMaster')).toHaveLength(1);
+        }
```
Comment (harelmo-lumigo) [inline test/instrumentations/mongodb/mongodb.test.ts:281]: Same as above about the branching ⬆️

**Review by harelmo-lumigo — APPROVED** [review]

File: src/samplers/mongodbSampler.ts (hunk 3 above)
Comment (eugene-lumigo) [reply src/samplers/mongodbSampler.ts:24]: agree

File: README.md (hunk 4 above)
Comment (eugene-lumigo) [reply README.md:126]: it was a request from Lir to make it a default behaviour

File: test/instrumentations/mongodb/mongodb.test.ts (hunk 6 above)
Comment (eugene-lumigo) [reply test/instrumentations/mongodb/mongodb.test.ts:245]: Done
